### ISSUE TITLE
Dot + Reduce Graph Fails to Compile on neuornx-cc 2.10

### ISSUE THREAD

**Issue by xanderdunn** [aws-neuron/aws-neuron-sdk#755]
We have a test graph with a simple dot product and reduce that's crashing the compiler. It's possible this is related to #754, the compiler output looks similar.

The graph: 
[rust_hlo_reduce.pb.zip](https://github.com/aws-neuron/aws-neuron-sdk/files/12743929/rust_hlo_reduce.pb.zip)

```
debug_ir: HloModule xla_computation_ordered_wrapper, entry_computation_layout={(f32[2,4]{1,0}, f32[4,8]{1,0}, f32[2]{0})->(f32[2]{0}, f32[2,4]{0,1})}

region_0.13 {
  Arg_0.10 = f32[] parameter(0)
  Arg_1.10 = f32[] parameter(1)
  ROOT add.12 = f32[] add(Arg_0.10, Arg_1.10)
}

ENTRY main.16 {
  Arg_0.1 = f32[2,4]{1,0} parameter(0)
  Arg_1.2 = f32[4,8]{1,0} parameter(1)
  dot.4 = f32[2,8]{0,1} dot(Arg_0.1, Arg_1.2), lhs_contracting_dims={1}, rhs_contracting_dims={0}
  constant.5 = f32[] constant(0)
  reduce.6 = f32[2]{0} reduce(dot.4, constant.5), dimensions={1}, to_apply=region_0.13
  Arg_2.3 = f32[2]{0} parameter(2)
  broadcast.7 = f32[2,8]{1,0} broadcast(Arg_2.3), dimensions={0}
  transpose.8 = f32[8,4]{0,1} transpose(Arg_1.2), dimensions={1,0}
  dot.9 = f32[2,4]{0,1} dot(broadcast.7, transpose.8), lhs_contracting_dims={1}, rhs_contracting_dims={0}
  ROOT tuple.14 = (f32[2]{0}, f32[2,4]{0,1}) tuple(reduce.6, dot.9)
}
```

The crash:
```
---- tensor::ops::tests::xla_reduce stdout ----
Python script stderr:
No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)

Running command 'neuronx-cc compile /mnt/drive1/tmp/rust_hlo_reduce.pb --framework=XLA --target=trn1 --model-type=transformer --auto-cast=none --output /mnt/drive1/tmp/reduce_15269776325709852363.neff'
Running command 'neuronx-cc compile /mnt/drive1/tmp/rust_hlo_reduce.pb --framework=XLA --target=trn1 --model-type=transformer --auto-cast=none --output /mnt/drive1/tmp/reduce_1966500338512357887.neff'
thread 'tensor::ops::tests::xla_reduce' panicked at 'Script '"neuronx-cc" "compile" "/mnt/drive1/tmp/rust_hlo_reduce.pb" "--framework=XLA" "--target=trn1" "--model-type=transformer" "--auto-cast=none" "--output" "/mnt/drive1/tmp/reduce_1966500338512357887.neff"' failed:
2023-09-27T22:49:20Z MM lhs/rhs par axes must be equal
, stdout:
2023-09-27T22:49:18Z Compilation is optimized for best performance and compilation time. For faster compilation time please use -O1
2023-09-27T22:49:18Z Running DoNothing
2023-09-27T22:49:19Z DoNothing finished after 0.004 seconds
2023-09-27T22:49:19Z Running CanonicalizeIR
2023-09-27T22:49:19Z CanonicalizeIR finished after 0.000 seconds
2023-09-27T22:49:19Z Running LegalizeCCOpLayout
2023-09-27T22:49:19Z LegalizeCCOpLayout finished after 0.000 seconds
2023-09-27T22:49:19Z Running ExpandBatchNorm
2023-09-27T22:49:19Z ExpandBatchNorm finished after 0.000 seconds
2023-09-27T22:49:19Z Running ResolveComplicatePredicates
2023-09-27T22:49:19Z ResolveComplicatePredicates finished after 0.000 seconds
2023-09-27T22:49:19Z Running AffinePredicateResolution
2023-09-27T22:49:19Z AffinePredicateResolution finished after 0.000 seconds
2023-09-27T22:49:19Z Running EliminateDivs
2023-09-27T22:49:19Z EliminateDivs finished after 0.000 seconds
2023-09-27T22:49:19Z Running PerfectLoopNest
2023-09-27T22:49:19Z PerfectLoopNest finished after 0.000 seconds
2023-09-27T22:49:19Z Running Simplifier
2023-09-27T22:49:19Z Simplifier finished after 0.006 seconds
2023-09-27T22:49:19Z Running GenericAccessSimplifier
2023-09-27T22:49:19Z GenericAccessSimplifier finished after 0.000 seconds
2023-09-27T22:49:19Z Running TCTransform
2023-09-27T22:49:19Z TCTransform finished after 0.001 seconds
2023-09-27T22:49:19Z Running CommuteConcat
2023-09-27T22:49:19Z CommuteConcat finished after 0.000 seconds
2023-09-27T22:49:19Z Running TensorOpFusion
2023-09-27T22:49:19Z TensorOpFusion finished after 0.000 seconds
2023-09-27T22:49:19Z Running TensorOpTransform
2023-09-27T22:49:19Z TensorOpTransform finished after 0.003 seconds
2023-09-27T22:49:19Z Running LowerTensorOp
2023-09-27T22:49:19Z LowerTensorOp finished after 0.000 seconds
2023-09-27T22:49:19Z Running CanonicalizeIR
2023-09-27T22:49:19Z CanonicalizeIR finished after 0.008 seconds
2023-09-27T22:49:19Z Running MemcpyElimination
2023-09-27T22:49:19Z MemcpyElimination finished after 0.004 seconds
2023-09-27T22:49:19Z Running LoopFusion
2023-09-27T22:49:19Z LoopFusion finished after 0.006 seconds
2023-09-27T22:49:19Z Running Simplifier
2023-09-27T22:49:19Z Simplifier finished after 0.001 seconds
2023-09-27T22:49:19Z Running Delinearization
2023-09-27T22:49:19Z Delinearization finished after 0.001 seconds
2023-09-27T22:49:19Z Running DeadStoreElimination
2023-09-27T22:49:19Z DeadStoreElimination finished after 0.009 seconds
2023-09-27T22:49:19Z Running Simplifier
2023-09-27T22:49:19Z Simplifier finished after 0.001 seconds
2023-09-27T22:49:19Z Running LICM
2023-09-27T22:49:19Z LICM finished after 0.000 seconds
2023-09-27T22:49:19Z Running Delinearization
2023-09-27T22:49:19Z Delinearization finished after 0.001 seconds
2023-09-27T22:49:19Z Running LoopFusion
2023-09-27T22:49:19Z LoopFusion finished after 0.001 seconds
2023-09-27T22:49:19Z Running SimplifySlice
2023-09-27T22:49:19Z SimplifySlice finished after 0.001 seconds
2023-09-27T22:49:19Z Running LICM
2023-09-27T22:49:19Z LICM finished after 0.002 seconds
2023-09-27T22:49:19Z Running Simplifier
2023-09-27T22:49:19Z Simplifier finished after 0.004 seconds
2023-09-27T22:49:19Z Running ValueNumbering
2023-09-27T22:49:19Z ValueNumbering finished after 0.003 seconds
2023-09-27T22:49:19Z Running LICM
2023-09-27T22:49:19Z LICM finished after 0.000 seconds
2023-09-27T22:49:19Z Running PadElimination
2023-09-27T22:49:19Z PadElimination finished after 0.000 seconds
2023-09-27T22:49:19Z Running Delinearization
2023-09-27T22:49:19Z Delinearization finished after 0.001 seconds
2023-09-27T22:49:19Z Running LoopFusion
2023-09-27T22:49:19Z LoopFusion finished after 0.002 seconds
2023-09-27T22:49:19Z Running GenericAccessSimplifier
2023-09-27T22:49:19Z GenericAccessSimplifier finished after 0.001 seconds
2023-09-27T22:49:19Z Running Simplifier
2023-09-27T22:49:19Z Simplifier finished after 0.006 seconds
2023-09-27T22:49:19Z Running LICM
2023-09-27T22:49:19Z LICM finished after 0.002 seconds
2023-09-27T22:49:19Z Running ValueNumbering
2023-09-27T22:49:20Z ValueNumbering finished after 0.002 seconds
2023-09-27T22:49:20Z Running TCTransform
2023-09-27T22:49:20Z TCTransform finished after 0.002 seconds
2023-09-27T22:49:20Z Running CommuteConcat
2023-09-27T22:49:20Z CommuteConcat finished after 0.001 seconds
2023-09-27T22:49:20Z Running RecognizeOpIdiom
2023-09-27T22:49:20Z RecognizeOpIdiom finished after 0.000 seconds
2023-09-27T22:49:20Z Running MaskPropagation
2023-09-27T22:49:20Z MaskPropagation finished after 0.001 seconds
2023-09-27T22:49:20Z Running Recompute
2023-09-27T22:49:20Z Recompute finished after 0.001 seconds
2023-09-27T22:49:20Z Running DeadCodeElimination
2023-09-27T22:49:20Z DeadCodeElimination finished after 0.000 seconds
2023-09-27T22:49:20Z Running DoNothing
2023-09-27T22:49:20Z DoNothing finished after 0.000 seconds
2023-09-27T22:49:20Z Running MutateDataType
2023-09-27T22:49:20Z MutateDataType finished after 0.000 seconds
2023-09-27T22:49:20Z Running AutoCastTCInputs
2023-09-27T22:49:20Z AutoCastTCInputs finished after 0.001 seconds
2023-09-27T22:49:20Z Running GenericAccessSimplifier
2023-09-27T22:49:20Z GenericAccessSimplifier finished after 0.000 seconds
2023-09-27T22:49:20Z Running Simplifier
2023-09-27T22:49:20Z Simplifier finished after 0.001 seconds
2023-09-27T22:49:20Z Running DelinearIndices
2023-09-27T22:49:20Z DelinearIndices finished after 0.002 seconds
2023-09-27T22:49:20Z Running Delinearization
2023-09-27T22:49:20Z Delinearization finished after 0.001 seconds
2023-09-27T22:49:20Z Running DelinearIndices
2023-09-27T22:49:20Z DelinearIndices finished after 0.002 seconds
2023-09-27T22:49:20Z Running DeadCodeElimination
2023-09-27T22:49:20Z DeadCodeElimination finished after 0.001 seconds
2023-09-27T22:49:20Z Running InferIntrinsicOnCC
2023-09-27T22:49:20Z InferIntrinsicOnCC finished after 0.003 seconds
2023-09-27T22:49:20Z Running ResolveAccessConflict
2023-09-27T22:49:20Z ResolveAccessConflict finished after 0.001 seconds
2023-09-27T22:49:20Z Running LICM
2023-09-27T22:49:20Z LICM finished after 0.003 seconds
2023-09-27T22:49:20Z Running LocalLayoutOpt
2023-09-27T22:49:20Z LocalLayoutOpt finished after 0.001 seconds
2023-09-27T22:49:20Z Running DelinearIndices
2023-09-27T22:49:20Z DelinearIndices finished after 0.001 seconds
2023-09-27T22:49:20Z Running PGLayoutTilingPipeline
2023-09-27T22:49:20Z Running PAGLayoutOpt
2023-09-27T22:49:20Z PAGLayoutOpt finished after 0.005 seconds
2023-09-27T22:49:20Z Running MaskPropagation
2023-09-27T22:49:20Z MaskPropagation finished after 0.001 seconds
2023-09-27T22:49:20Z Running CanonicalizeDAGForPGTiling
2023-09-27T22:49:20Z CanonicalizeDAGForPGTiling finished after 0.001 seconds
2023-09-27T22:49:20Z Running PGTiling
2023-09-27T22:49:20Z PGTiling finished after 0.000 seconds
2023-09-27T22:49:20Z PGLayoutTilingPipeline finished after 0.054 seconds
2023-09-27T22:49:20Z
2023-09-27T22:49:20Z Diagnostic information:
2023-09-27T22:49:20Z   NeuronX Compiler version 2.10.0.34+6c8792c6f
2023-09-27T22:49:20Z
2023-09-27T22:49:20Z   Python version 3.8.10
2023-09-27T22:49:20Z   HWM version 2.10.0.5-7b1976adf
2023-09-27T22:49:20Z   NumPy version 1.21.6
2023-09-27T22:49:20Z
2023-09-27T22:49:20Z   Running on AMI ami-01257e71ecb2f431c
2023-09-27T22:49:20Z   Running in region usw2-az4
2023-09-27T22:49:20Z
2023-09-27T22:49:20Z Diagnostic logs stored in /home/ubuntu/dev/Kholinar/xla/log-neuron-cc.txt
', xla/src/xla_utils.rs:49:5
```

We're running:
```
$ neuronx-cc --version
NeuronX Compiler version 2.10.0.35+3817a0c8c

Python version 3.8.10
HWM version 2.10.0.5-7b1976adf
NumPy version 1.21.6

Running on AMI ami-01257e71ecb2f431c
Running in region usw2-az4
```

This successfully compiled and ran on previous versions of neuronx-cc.

**Comment by awsilya** [comment]
@xanderdunn  ... looking

**Comment by awsilya** [comment]
we will try to reproduce and get back to you.  Same applies to the other 2 tickets.

**Comment by aws-donkrets** [comment]
Hi [xanderdunn](https://github.com/xanderdunn), we have been able to reproduce the issue and are testing a fix. We discovered it too late for our upcoming (2.15) release but expect it to be available in a following release.

**Comment by xanderdunn** [comment]
Thanks @aws-donkrets!

**Comment by xanderdunn** [comment]
This reduce is fixed as of `neuronx_cc-2.12.54.0+f631c2365` Thank you! :)
I'll close this when it reaches release.

**Comment by xanderdunn** [comment]
This is fixed in 2.16 release
